app/models/robot/pick_data.rb

```diff
@@ -6,18 +6,24 @@ class Robot::PickData
 
   delegate :requests, to: :batch
 
-  def initialize(batch, target_barcode, user: batch.user)
+  def initialize(batch, target_barcode, user: batch.user, max_beds: nil)
     @batch = batch
```

Comment:
The robots with no beds configured are passing in zero to this method, not nil (as nil.to_i is zero, not nil). Further down you treat nil as though it's a robot with infinite beds. If we go for that behaviour, we might be better setting the default to `Float::INFINITY` here, and that way we wouldn't even need to have the conditional in generate_picking_data_hash.

However, given almost everywhere this is used then expects the robots to have matching named beds configures, we're probably just punting the problem further down the line. It might be best just to keep the default max beds as zero, and throw an exception if we see it. At the moment I think we're at risk on an infinite loop.

Background: The robots without max_beds configured aren't Tecans/Hamiltons and haven't been used for cherrypicking. They are in the database purely for maintaining unique barcodes etc. Recently we added a scope to filter then out, so should probably make sure we use that in the dropdown list in the pipeline, and add validation in the robot verification process to make sure they can'r be used.

Reply:
Thanks for the thoughts. 

I set it to nil and put this behaviour in as the easiest way to keep the existing `expected_layout` method functioning as before, with the expectation that we'd get rid of it and those existing methods once we'd done all the cherrypicking stories.

Can I let you change it as part of your story, as once that's done we can probably get rid of `expected_layout` and related methods?

P.S. I don't think we're at risk of an infinite loop - how?